### PR TITLE
Implement admission controller

### PR DESCRIPTION
**Pull request by tkulczynski**
Fixing #534. 
Tested e2e: I run a cluster with a single VPA that had recommendation set and it used it when new nodes were created.

### CODE REVIEW

**Comment by tkulczynski** [comment]
cc @mwielgus

File: vertical-pod-autoscaler/admission-controller/config.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		Certificates: []tls.Certificate{sCert},
+		ClientCAs:    apiserverCA,
+		ClientAuth:   tls.NoClientCert, //RequireAndVerifyClientCert,
+	}
+}
+
+// register this example webhook admission controller with the kube-apiserver
+// by creating externalAdmissionHookConfigurations.
+func selfRegistration(clientset *kubernetes.Clientset, caCert []byte) {
+	time.Sleep(10 * time.Second)
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/config.go]: Why is this needed?

File: vertical-pod-autoscaler/admission-controller/config.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+			glog.Fatal(err2)
+		}
+	}
+	webhookConfig := &v1beta1.MutatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: webhookConfigName,
+		},
+		Webhooks: []v1beta1.Webhook{
+			{
+				Name: "hpa.k8s.io",
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/config.go]: hpa?

File: vertical-pod-autoscaler/admission-controller/config.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+					Rule: v1beta1.Rule{
+						APIGroups:   []string{""},
+						APIVersions: []string{"v1"},
+						Resources:   []string{"pods"},
+					},
+				}},
+				ClientConfig: v1beta1.WebhookClientConfig{
+					Service: &v1beta1.ServiceReference{
+						Namespace: "kube-system",
+						Name:      "hpa-webhook",
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/config.go]: s/hpa/vpa here and elsewhere.

File: vertical-pod-autoscaler/admission-controller/gencerts.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Generates the a CA cert, a server key, and a server cert signed by the CA.
+# reference:
+https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/webhook/gencerts.sh
+set -e
+
+CN_BASE="hpa_webhook"
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/gencerts.sh]: vpa

File: vertical-pod-autoscaler/admission-controller/gencerts.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+*/
+EOF
+
+echo "// This file was generated using openssl by the gencerts.sh script" >> $outfile
+echo "" >> $outfile
+echo "package main" >> $outfile
+for file in caKey caCert serverKey serverCert; do
+  data=$(cat ${file}.pem)
+  echo "" >> $outfile
+  echo "var $file = []byte(\`$data\`)" >> $outfile
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/gencerts.sh]: Embedding certs in image/code is a bad idea. The suggested way of handling certs in K8S is through secrets mechanism. If you are ok with changing certs with every pod restart you can also generate them inside the docker image as a part of the startup script.

File: vertical-pod-autoscaler/admission-controller/certs.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file was generated using openssl by the gencerts.sh script
+
+package main
+
+var caKey = []byte(`-----BEGIN RSA PRIVATE KEY-----
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/certs.go]: There is no point in any security here if we are publishing private certs on github :).

File: vertical-pod-autoscaler/admission-controller/config.go (hunk 1 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/config.go]: Not sure what you mean. Registration is needed so that the api server knows there's an admission controller out there. It's done here not in deploy/*.yaml, because it needs the certs.

File: vertical-pod-autoscaler/admission-controller/config.go (hunk 2 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/config.go]: Done. Gosh, sorry for that, I don't know what I was thinking about.

File: vertical-pod-autoscaler/admission-controller/config.go (hunk 3 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/config.go]: Done

File: vertical-pod-autoscaler/admission-controller/gencerts.sh (hunk 4 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/gencerts.sh]: Done

File: vertical-pod-autoscaler/admission-controller/gencerts.sh (hunk 5 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/gencerts.sh]: Went with secrets. The embedded certs were part of the admission webhook example that you pointed me to, but of course you're correct it's not a great production code ;)

File: vertical-pod-autoscaler/admission-controller/certs.go (hunk 6 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/certs.go]: Done.

**Comment by tkulczynski** [comment]
@mwielgus PTAL

File: vertical-pod-autoscaler/admission-controller/certs.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		glog.Error(err)
+		return nil
+	}
+	glog.Infof("Successfuly read %d bytes from %v", count, filePath)
+	return res
+}
+
+func initCerts(certsDir string) {
+	caKey = readFile(path.Join(certsDir, "caKey.pem"))
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/certs.go]: make it const.

File: vertical-pod-autoscaler/admission-controller/certs.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+import (
+	"os"
+	"path"
+
+	"github.com/golang/glog"
+)
+
+var (
+	caKey, caCert, serverKey, serverCert []byte
```
Comment (mwielgus) [inline vertical-pod-autoscaler/admission-controller/certs.go]: why do we need global values for these?

Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/certs.go]: Well, I was just doing minimal changes and I probably don't feel go style yet.

File: vertical-pod-autoscaler/admission-controller/certs.go (hunk 7 above)
Comment (tkulczynski) [reply vertical-pod-autoscaler/admission-controller/certs.go]: I think this comment is obsoleted by the other one, or did I misunderstand?
